modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizerSelfTest.java

```diff
@@ -0,0 +1,591 @@
+        String subSql = "select id from dep";
+
+        String resSql = String.format(outerSqlTemplate, subSql);
+
+        check(resSql, 1);
+    }
+
+    /**
+     * Simple case, but tables are the same.
+     */
+    @Test
+    public void testInClause5() {
+        String outerSqlTemplate = "select id, name from emp e where id in (%s)";
+        String subSql = "select id from emp";
+
+        String resSql = String.format(outerSqlTemplate, subSql);
+
+        check(resSql, 1);
+    }
+
+    /**
+     * Simple case, but inner table has coumpound PK.
+     */
+    @Test
+    public void testInClauseCompoundPk() {
+        String outerSqlTemplate = "select id, name from emp e where id in (%s)";
+        String subSql = "select id from dep2 where id2 = 12";
+
+        String resSql = String.format(outerSqlTemplate, subSql);
+
+        check(resSql, 1);
+    }
+
+    /**
+     * Case with multiple IN clauses.
+     */
+    @Test
+    public void testInClauseMultiple() {
+        String outerSqlTemplate = "select id, name from dep2 d where id in (%s) and id2 in (%s)";
+        String subSql = "select id from emp";
+
+        String resSql = String.format(outerSqlTemplate, subSql, subSql);
+
+        check(resSql, 1);
+    }
+
+    /**
+     * Test should verify all cases where subquery should not be rewrited.
+     */
+    @Test
+    public void testOptimizationShouldNotBeApplied() {
+        String sql = "" +
+            // follow should not be rewrited beacuse of aggregates
+            "select (select max(id) from emp) f1," +
+            "       (select sum(id) from emp) f2," +
+            "       (select distinct id from emp where id = 1) f3," +
+            "       (select distinct(id) from emp where id = 1) f4," +
+            "       (select id from emp where id = 1 group by id) f5," +
+            "       (select id from emp where id = 1 group by 1) f6," +
+            "       (select id from emp limit 1) f7," +
+            "       (select id from emp where id = 2 offset 2) f8," +
+            // and this one because dep2 has compound pk (id, id2),
+            // so predicate over 'id' could not guarantee uniqueness
+            "       (select id from dep2 where id = 2) f9" +
+            " from dep";
+
+        check(sql, 10);
+    }
+
+    /**
+     * @param sql Sql.
+     * @param expSelectClauses Expected select clauses.
+     */
+    private void check(String sql, int expSelectClauses) {
+        optimizationEnabled(false);
+
+        List<List<?>> exp = cache.query(new SqlFieldsQuery(sql)).getAll();
+
+        optimizationEnabled(true);
+
+        List<List<?>> act = cache.query(new SqlFieldsQuery(sql).setEnforceJoinOrder(true)).getAll();
+
+        Assert.assertEquals("Result set mismatch", exp, act);
+
+        String plan = cache.query(new SqlFieldsQuery("explain " + sql)).getAll().get(0).get(0).toString();
+
+        System.out.println(plan);
```

Comment:
println

Reply:
think it's OK to have println in the tests. It serves debugging purpose.